Forge.TreeWalker/src/TreeNodeContext.cs

```diff
@@ -0,0 +1,100 @@
+        /// </summary>
+        public Guid SessionId { get; private set; }
+
+        /// <summary>
+        /// The tree node key where the Action resides.
+        /// </summary>
+        public string TreeNodeKey { get; private set; }
+
+        /// <summary>
+        /// The dynamic properties of this Node given by the ForgeTree schema.
+        /// </summary>
+        public dynamic Properties { get; private set; }
+
+        /// <summary>
+        /// The dynamic user-defined context object that is able to be referenced when evaluating schema expressions or performing actions.
+        /// </summary>
+        public object UserContext { get; private set; }
+
+        /// <summary>
+        /// The cancellation token.
+        /// </summary>
+        public CancellationToken Token { get; private set; }
+
+        /// <summary>
+        /// The name of the ForgeTree in the JsonSchema.
+        /// </summary>
+        public string TreeName { get; private set; }
+
+        /// <summary>
+        /// The unique identifier for the root/parent tree walking session.
+        /// </summary>
+        public Guid RootSessionId { get; private set; }
+
+        /// <summary>
+        /// Indicates whether the tree walker should skip all actions defined in this node, and proceed with ChildSelector.
+        /// </summary>
+        /// <remarks>Allow Public set for testing purpose.</remarks>
+        public bool ShouldSkipActionsInTreeNode { get; set; }
+
+        /// <summary>
+        /// Instantiates an TreeNodeContext object.
+        /// </summary>
+        /// <param name="sessionId">The unique identifier for this tree walking session.</param>
+        /// <param name="treeNodeKey">The TreeNode's key where the Action is taking place.</param>
+        /// <param name="properties">The properties of this Tree node..</param>
+        /// <param name="userContext">The user context for this Action.</param>
+        /// <param name="token">The cancellation token.</param>
+        /// <param name="treeName">The name of the ForgeTree in the JsonSchema.</param>
+        /// <param name="rootSessionId">The unique identifier for the root/parent tree walking session.</param>
+        /// <param name="ShouldSkipActionsInTreeNode">Indicates whether the tree walker should skip all actions defined in this node, and proceed with ChildSelector</param>
+        public TreeNodeContext(
+            Guid sessionId,
+            string treeNodeKey,
+            dynamic properties,
+            object userContext,
+            CancellationToken token,
+            string treeName,
+            Guid rootSessionId,
+            bool shouldSkipActionsInTreeNode)
+        {
+            if (sessionId == null) throw new ArgumentNullException("sessionId");
+            if (string.IsNullOrWhiteSpace(treeNodeKey)) throw new ArgumentNullException("treeNodeKey");
+            if (userContext == null) throw new ArgumentNullException("userContext");
```

Comment:
>userContext [](start = 16, length = 11)

userContext can be null. It's an optional parameter.

Reply:
removed the null check, and added the comment
userContext is an optional parameter for TreeNode, so we don't throw ArgumentNullException when null.

---
In reply to: [504159484](https://github.com/microsoft/Forge/pull/32#discussion_r504159484) [](ancestors = 504159484)